crates/wiggle/generate/src/module_trait.rs

```diff
@@ -82,9 +82,9 @@ pub fn define_module_trait(m: &Module, settings: &CodegenSettings) -> TokenStrea
         };
 
         if is_anonymous {
-            quote!(#asyncness fn #funcname(&mut self, #(#args),*) -> #result; )
+            quote!(#asyncness fn #funcname(&self, #(#args),*) -> #result; )
         } else {
-            quote!(#asyncness fn #funcname<#lifetime>(&mut self, #(#args),*) -> #result;)
+            quote!(#asyncness fn #funcname<#lifetime>(&self, #(#args),*) -> #result;)
         }
```

Comment:
Wiggle is used for more than just WASI, and I don't think it's suitable to make this change. Providing `&mut` access shows how Wasmtime gives this ability even if WASI doesn't take advantage of it.

Instead the traits should still give access to `&mut self` but `WasiCtx` would internally hold `Arc<Table>` or something like that which is where the requirement for locking comes in (because `&mut Arc<T>` still only gives you `&T`)

Reply:
`Arc<T>` sounds like a plan. Thanks!

Reply:
What do you think about:
```rust
impl wasi_snapshot_preview1::WasiSnapshotPreview1 for Arc<WasiCtx>
```

Reply:
I would personally prefer to put the `Arc` inside of the `WasiCtx` to make it more ergonomic to use but either way achieves the same goal.

Reply:
deferred this for a later PR